buildbot_nix/buildbot_nix/gitlab_project.py

```diff
@@ -0,0 +1,317 @@
+            ReloadGitlabProjects(self.config, self.config.project_cache_file),
+        )
+        factory.addStep(
+            CreateGitlabProjectHooks(
+                self.config,
+                self.instance_url,
+            )
+        )
+        return util.BuilderConfig(
+            name=self.reload_builder_name,
+            workernames=worker_names,
+            factory=factory,
+        )
+
+    def create_reporter(self) -> ReporterBase:
+        return GitLabStatusPush(
+            token=self.config.token,
+            context=Interpolate("buildbot/%(prop:status_name)s"),
+            baseURL=self.config.instance_url,
+            generators=[
+                BuildNixEvalStatusGenerator(),
+            ],
+        )
+
+    def create_change_hook(self) -> dict[str, Any]:
+        return dict(secret=self.config.webhook_secret)
+
+    def load_projects(self) -> list["GitProject"]:
+        if not self.config.project_cache_file.exists():
+            return []
+
+        repos: list[RepoData] = filter_repos_by_topic(
+            self.config.topic,
+            sorted(
+                model_validate_project_cache(RepoData, self.config.project_cache_file),
+                key=lambda repo: repo.path_with_namespace,
+            ),
+            lambda repo: repo.topics,
+        )
+        tlog.info(f"Loading {len(repos)} cached repos.")
+
+        return [
+            GitlabProject(self.config, RepoData.model_validate(repo)) for repo in repos
+        ]
+
+    def are_projects_cached(self) -> bool:
+        return self.config.project_cache_file.exists()
+
+    def create_auth(self) -> AuthBase:
+        raise NotImplementedError
+
+    def create_avatar_method(self) -> AvatarBase | None:
+        return None
+
+    @property
+    def reload_builder_name(self) -> str:
+        return "reload-gitlab-projects"
+
+    @property
+    def type(self) -> str:
+        return "gitlab"
+
+    @property
+    def pretty_type(self) -> str:
+        return "Gitlab"
+
+    @property
+    def change_hook_name(self) -> str:
+        return "gitlab"
+
+
+class ReloadGitlabProjects(ThreadDeferredBuildStep):
+    name = "reload_gitlab_projects"
+
+    config: GitlabConfig
+    project_cache_file: Path
+
+    def __init__(
+        self,
+        config: GitlabConfig,
+        project_cache_file: Path,
+        **kwargs: Any,
+    ) -> None:
+        self.config = config
+        self.project_cache_file = project_cache_file
+        super().__init__(**kwargs)
+
+    def run_deferred(self) -> None:
+        repos: list[RepoData] = filter_repos_by_topic(
+            self.config.topic,
+            refresh_projects(self.config, self.project_cache_file),
+            lambda repo: repo.topics,
+        )
+        atomic_write_file(self.project_cache_file, model_dump_project_cache(repos))
+
+    def run_post(self) -> Any:
+        return util.SUCCESS
+
+
+class CreateGitlabProjectHooks(ThreadDeferredBuildStep):
+    name = "create_gitlab_project_hooks"
```

Comment:
Btw. if gitlab has a different way to install webhooks i.e. through some app interface, this might be preferred. It's a bit of a historical accident in github that we didn't just use the webhooks that are provided by a github app rather than installing an explicit webhook into each repository.